pallets/amm/src/pool_structs.rs

```diff
@@ -0,0 +1,161 @@
+// Copyright 2021 Parallel Finance Developer.
+// This file is part of Parallel Finance.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Groups common pool related structures
+
+use parallel_primitives::{Balance, CurrencyId, Rate};
+use sp_runtime::{traits::Saturating, ArithmeticError, DispatchError, FixedPointNumber};
+
+// Amplification Coefficient Weight.
+//
+// In this pallet, the actual amplification coefficient will be `exchange_rate` * `ACW`.
+const ACW: Rate = Rate::from_inner(Rate::DIV / 100 * 50); // 50%
+
+#[derive(
+    Clone,
+    PartialEq,
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    sp_runtime::RuntimeDebug,
+)]
+pub enum SwapType {
+    Buy,
+    Sell,
+}
+
+// Wrapper around the result of `Pallet::calculate_amount`
+pub struct AmountEvaluation {
+    pub account_amount: Balance,
+    pub pool_amount: Balance,
+}
+
+#[derive(
+    Clone,
+    PartialEq,
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    sp_runtime::RuntimeDebug,
+)]
+pub struct LiquidityProviderAmounts {
+    pub base_amount: Balance,
+    pub quote_amount: Balance,
+}
+
+#[derive(
+    Clone,
+    PartialEq,
+    parity_scale_codec::Decode,
+    parity_scale_codec::Encode,
+    sp_runtime::RuntimeDebug,
+)]
+pub struct Pool {
+    pub base_amount: Balance,
+    pub quote_amount: Balance,
+    pub base_asset: CurrencyId,
+    pub quote_asset: CurrencyId,
+}
+
+pub struct StandardSwap;
+
+pub struct StableSwap;
+
+pub struct StakingSwap;
+
+pub trait StabilityPool {}
+
+impl StabilityPool for () {}
+
+pub trait AMMCurve {
+    // Calculates the amount according to the underlying formula and the provided pool.
+    fn calculate_amount(
+        exchange_rate: Rate,
+        new_amount: Balance,
+        pool: &Pool,
+    ) -> Result<AmountEvaluation, DispatchError>;
+}
```

Comment:
To avoid relying on native runtime primitives :)

```suggestion
pub trait AMMCurve<ExchangeRate, Balance> {
    // Calculates the amount according to the underlying formula and the provided pool.
    fn calculate_amount(
        exchange_rate: ExchangeRate,
        new_amount: Balance,
        pool: &Pool,
    ) -> Result<AmountEvaluation, DispatchError>;
}
```

Reply:
Same as https://github.com/parallel-finance/parallel/pull/365#discussion_r686344879